### PR TITLE
inference: add reflection utility for exception type analysis

### PR DESCRIPTION
**Pull request by aviatesk**
This commit defines functions that mirror our tools for analyzing return types and computational effects.

The key point to discuss is that this commit introduces two functions: `Base.infer_exception_types` and `Base.infer_exception_type`. `Base.infer_exception_types` acts like `Base.return_types`, giving a list of exception types for each method that matches with the given call signature. On the other hand, `Base.infer_exception_type` is akin to `Base.infer_effects`, returning a single exception type that covers all potential outcomes entailed by the given call signature. I personally lean towards the latter for its utility, particularly in testing scenarios, but I included `infer_exception_types` too for consistency with `return_types`. I'd welcome any feedback on this approach.

### CODE REVIEW

**Comment by Keno** [comment]
> Base.exception_type

I think this name is potentially confusing and should probably use an `infer_` name like `infer_effects`.

File: base/reflection.jl [hunk 1]
```diff
@@ -1525,7 +1525,7 @@ internals.
 - `world::UInt = Base.get_world_counter()`: optional, controls the world age to use
   when looking up methods, use current world age if not specified.
 - `interp::Core.Compiler.AbstractInterpreter = Core.Compiler.NativeInterpreter(world)`:
-  optional, controls the abstract interpreter to use, use the native interpreter if not specified.
+  optional, controls the abstract interpreter to use, use The abstract interpreter if not specified.
```
Comment (Keno) [inline base/reflection.jl:1528]: I don't understand the intention of this change.

**Review by Keno — APPROVED** [review]
LGTM modulo the two comments I made.

File: base/reflection.jl (hunk 1 above)
Comment (aviatesk) [reply base/reflection.jl]: Thanks, remnants of find and replace.

**Pull request merged**